### PR TITLE
Update C++ dependencies, roughly matching ones used in python.

### PR DESCRIPTION
**Pull request by jcferretti**
The goal is actually to build python dependencies with debug info; this is a step.

The change to base-images happened first here: https://github.com/deephaven/deephaven-base-images/pull/116

### CODE REVIEW

**Review by kosak — APPROVED** [review]

File: docker/registry/cpp-clients-multi-base/gradle.properties [hunk 1]
```diff
@@ -1,5 +1,4 @@
 io.deephaven.project.ProjectType=DOCKER_REGISTRY
 deephaven.registry.imageName=ghcr.io/deephaven/cpp-clients-multi-base:latest
-# When the image version changes, remember to update cpp-client/build-dependencies.sh to match the image's content
```
Comment (devinrsmith) [inline docker/registry/cpp-clients-multi-base/gradle.properties:3]: Leave this comment?

**Comment by devinrsmith** [comment]
>  Execution failed for task ':cpp-client:compareProtobuf'.
> Sources do not match expected, re-run :cpp-client:updateProtobuf

**Comment by devinrsmith** [comment]
I think the version of protobuf will need to be updated in concert with the version grpc supports ie #5299

```
[Task :proto:proto-backplane-grpc:compileJava FAILED] cannot access GeneratedMessageV3
                Status.newBuilder().setCode(statusCode.getNumber()).setMessage(details).build());
                      ^
  class file for com.google.protobuf.GeneratedMessageV3 not found
```
  
```
./gradlew -q proto:proto-backplane-grpc:dependencyInsight --configuration compileClasspath --dependency proto-google-common-protos 
com.google.api.grpc:proto-google-common-protos:2.22.0
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 8            |

com.google.api.grpc:proto-google-common-protos:2.22.0
\--- io.grpc:grpc-protobuf:1.58.0
     +--- compileClasspath (requested io.grpc:grpc-protobuf)
     +--- org.apache.arrow:flight-core:13.0.0 (requested io.grpc:grpc-protobuf:1.56.0)
     |    \--- compileClasspath
     \--- io.grpc:grpc-bom:1.58.0
          \--- compileClasspath
```

**Comment by devinrsmith** [comment]
https://github.com/grpc/grpc-java/issues/11015

**Comment by devinrsmith** [comment]
Looks like grpc-java is sticking with 3.x at this time due to known issues. The advice right now is
> For those watching at home and want to be prepared, use protobuf-java 3.25 and protoc 25.

https://github.com/protocolbuffers/protobuf/issues/16452#issuecomment-2076360676

**Review by chipkent — APPROVED** [review]

**Review by chipkent — APPROVED** [review]

**Review by kosak — APPROVED** [review]

**Review by devinrsmith — APPROVED** [review]

File: cpp-client/build-dependencies.sh [hunk 2]
```diff
@@ -530,8 +531,8 @@ if [ "$CLONE_ZLIB" = "yes" ]; then
   echo
   echo "*** Clone zlib"
   cd $SRC
-  # Previously used version: v1.2.11
-  git clone $GIT_FLAGS -b v1.2.13 --depth 1 "${GITHUB_BASE_URL}/madler/zlib"
+  # Previously used version: v1.2.13
+  git clone $GIT_FLAGS -b v1.3 --depth 1 "${GITHUB_BASE_URL}/madler/zlib"
```
Comment (chipkent) [inline cpp-client/build-dependencies.sh:535]: Do we need to specify the patch version, or are we ok with that changing?

File: cpp-client/build-dependencies.sh [hunk 3]
```diff
@@ -561,24 +562,25 @@ if [ "$CLONE_PROTOBUF" = "yes" ]; then
   echo
   echo "*** Cloning protobuf"
   cd $SRC
-  # Previously used version: v3.21.12
-  git clone $GIT_FLAGS -b v3.21.2 --depth 1 "${GITHUB_BASE_URL}/protocolbuffers/protobuf.git"
+  # Previously used version: v3.21.2
+  git clone $GIT_FLAGS -b v25.3 --depth 1 "${GITHUB_BASE_URL}/protocolbuffers/protobuf.git"
```
Comment (chipkent) [inline cpp-client/build-dependencies.sh:566]: 1. Were we ultra out of date?
2. Did they change their versioning convention?
3. Do we need to specify the patch version?

**Review by alexpeters1208 — APPROVED** [review]
R stuff looks good

File: cpp-client/build-dependencies.sh (hunk 2 above)
Comment (jcferretti) [reply cpp-client/build-dependencies.sh:535]: That's a specific version, is just the name they gave, in git terms is just some arbitrary string.  Mean as far as I understand how the universe works, that won't change when they create a new version.

**Review by chipkent — APPROVED** [review]

**Pull request merged**